File: prototype/sqlite_import.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
SQLite database import functionality for ProllyTree.

Provides functions for importing SQLite databases into ProllyTree stores.
"""

import sqlite3
import time
from typing import Optional, List

from db import DB
from store import Store, CachedFSStore
from cursor import TreeCursor


def validate_tree_sorted(db: DB, table_name: str) -> bool:
    """
    Validate that all keys in the tree are in sorted order with no duplicates.

    Args:
        db: DB instance
        table_name: Name of table being validated (for error messages)

    Returns:
        True if valid, False if duplicates or unsorted keys found
    """
    store = db.get_store()
    root_hash = db.get_root_hash()

    cursor = TreeCursor(store, root_hash)
    prev_key = None
    position = 0
    duplicates = []
    unsorted = []

    entry = cursor.next()
    while entry:
        position += 1
        key = entry[0]

        if prev_key is not None:
            if key == prev_key:
                duplicates.append((position, key))
                if len(duplicates) <= 5:  # Show first 5
                    print(f"  DUPLICATE at position {position}: {key}")
            elif key < prev_key:
                unsorted.append((position, prev_key, key))
                if len(unsorted) <= 5:  # Show first 5
                    print(f"  UNSORTED at position {position}: {prev_key} > {key}")

        prev_key = key
        entry = cursor.next()

    if duplicates or unsorted:
        print(f"\nVALIDATION FAILED for {table_name}:")
        print(f"  Total entries: {position:,}")
        print(f"  Duplicates found: {len(duplicates)}")
        print(f"  Unsorted pairs found: {len(unsorted)}")
        return False

    print(f"  Validation: {position:,} keys, all sorted, no duplicates ✓")
    return True


def import_sqlite_table(db: DB, sqlite_conn: sqlite3.Connection, table_name: str,
                        batch_size: int = 1000, verbose_batches: bool = False,
                        validate: bool = False) -> int:
    """
    Import a single SQLite table into the database.

    Args:
        db: DB instance
        sqlite_conn: SQLite connection
        table_name: Name of table to import
        batch_size: Batch size for inserts
        verbose_batches: Show detailed batch statistics
        validate: Validate tree is sorted after import

    Returns:
        Number of rows imported
    """
    cursor = sqlite_conn.cursor()

    # Get primary key columns
    cursor.execute(f"PRAGMA table_info({table_name})")
    rows = cursor.fetchall()

    pk_columns = []
    for row in rows:
        if row[5]:  # pk column is at index 5
            pk_columns.append((row[5], row[1]))  # (pk_position, column_name)

    if pk_columns:
        pk_columns.sort(key=lambda x: x[0])
        primary_key = [col[1] for col in pk_columns]
    else:
        primary_key = ["rowid"]

    # Get column names and types
    cursor.execute(f"PRAGMA table_info({table_name})")
    table_info = cursor.fetchall()
    columns = [row[1] for row in table_info]
    column_types = [row[2] for row in table_info]

    # Display table info
    if len(primary_key) == 1:
        print(f"Primary key: {primary_key[0]}")
    else:
        print(f"Primary key (compound): {', '.join(primary_key)}")

    # Get row count
    cursor.execute(f"SELECT COUNT(*) FROM {table_name}")
    row_count = cursor.fetchone()[0]
    print(f"Total rows: {row_count:,}")

    if row_count == 0:
        print("Skipping empty table")
        return 0

    # Create table in DB
    print(f"Stored schema: {len(columns)} columns")
    db.create_table(table_name, columns, column_types, primary_key)

    # Prepare row iterator - no ORDER BY, we'll sort each batch in Python
    if primary_key == ["rowid"]:
        cursor.execute(f"SELECT rowid, * FROM {table_name}")
    else:
        cursor.execute(f"SELECT * FROM {table_name}")

    def row_generator():
        """Generator that yields rows from cursor."""
        while True:
            rows = cursor.fetchmany(batch_size)
            if not rows:
                break

            for row in rows:
                if primary_key == ["rowid"]:
                    # Skip rowid in the values, it's handled by insert_rows
                    yield row
                else:
                    yield row

    # Insert rows
    table_start = time.time()

    # Enable per-batch validation if requested
    if validate:
        db._validate_after_batch = True

    rows_processed = db.insert_rows(table_name, row_generator(),
                                    batch_size=batch_size,
                                    verbose=verbose_batches)

    # Disable per-batch validation
    if validate:
        db._validate_after_batch = False

    table_time = time.time() - table_start
    table_rate = rows_processed / table_time if table_time > 0 else 0

    print(f"\nCompleted {table_name}: {rows_processed:,} rows in {table_time:.2f}s "
          f"({table_rate:,.0f} rows/sec)")

    # Print root hash
    root_hash = db.get_root_hash()
    print(f"  Root hash: {root_hash}")

    # Validate if requested
    if validate:
        print(f"\nValidating tree for {table_name}...")
        is_valid = validate_tree_sorted(db, table_name)
        if not is_valid:
            raise ValueError(f"Tree validation failed for {table_name} - contains duplicates or unsorted keys!")

    # Show cache stats if available
    store = db.get_store()
    if isinstance(store, CachedFSStore):
        creation_stats = store.get_creation_stats()
        cache_stats = store.get_cache_stats()
        print(f"  Cumulative: {creation_stats['total_leaves_created']:,} leaves, "
              f"{creation_stats['total_internals_created']:,} internals created")
        print(f"  Cache: {cache_stats['cache_evictions']:,} evictions, "
              f"{cache_stats['cache_size']:,}/{cache_stats['max_cache_size']:,} entries, "
              f"{cache_stats['hit_rate']} hit rate")

        # Print size distributions
        print()
        store.print_distributions(bucket_count=10)

    return rows_processed


def import_sqlite_database(db_path: str, store: Store,
                           pattern: float = 0.0001, seed: int = 42,
                           batch_size: int = 1000,
                           tables_filter: Optional[List[str]] = None,
                           verbose_batches: bool = False,
                           validate: bool = False) -> DB:
    """
    Import a SQLite database into ProllyTree.

    Args:
        db_path: Path to SQLite database
        store: Store instance to use
        pattern: ProllyTree split pattern
        seed: Random seed
        batch_size: Batch size for inserts
        tables_filter: Optional list of table names to import
        verbose_batches: Show detailed batch statistics
        validate: Validate tree is sorted after each table import

    Returns:
        DB instance
    """
    print(f"Opening database: {db_path}")
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    # Get all table names
    cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'")
    all_tables = [row[0] for row in cursor.fetchall()]

    # Filter tables if requested
    if tables_filter:
        tables = [t for t in all_tables if t in tables_filter]
        print(f"Found {len(all_tables)} tables, importing {len(tables)}: {', '.join(tables)}")
        skipped = [t for t in tables_filter if t not in all_tables]
        if skipped:
            print(f"Warning: Requested tables not found: {', '.join(skipped)}")
    else:
        tables = all_tables
        print(f"Found {len(tables)} tables: {', '.join(tables)}")

    # Create DB
    print(f"\nInitializing ProllyTree (pattern={pattern}, seed={seed})")
    if isinstance(store, CachedFSStore):
        cache_stats = store.get_cache_stats()
        print(f"  Cache size: {cache_stats['max_cache_size']}")

    db = DB(store=store, pattern=pattern, seed=seed, validate=validate)

    # Import each table
    total_rows = 0
    total_start = time.time()

    for table_name in tables:
        print(f"\n{'='*80}")
        print(f"Importing table: {table_name}")
        print(f"{'='*80}")

        rows_imported = import_sqlite_table(db, conn, table_name,
                                            batch_size=batch_size,
                                            verbose_batches=verbose_batches,
                                            validate=validate)
        total_rows += rows_imported

    total_time = time.time() - total_start
    total_rate = total_rows / total_time if total_time > 0 else 0

    # Get final root hash
    final_root_hash = db.get_root_hash()

    print(f"\n{'='*80}")
    print(f"IMPORT COMPLETE")
    print(f"{'='*80}")
    print(f"Total rows imported: {total_rows:,}")
    print(f"Total time: {total_time:.2f}s")
    print(f"Overall rate: {total_rate:,.0f} rows/sec")
    print(f"\n{'='*80}")
    print(f"FINAL ROOT HASH: {final_root_hash}")
    print(f"{'='*80}")

    # Print final stats
    print(f"\nTree statistics:")
    print(f"  Store type: {type(store).__name__}")
    print(f"  Total nodes in storage: {store.count_nodes():,}")

    if isinstance(store, CachedFSStore):
        stats = store.get_cache_stats()
        print(f"\nCache statistics:")
        for key, value in stats.items():
            print(f"  {key}: {value}")

    conn.close()
    return db

File: examples/sqlite_import.rs
/*
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! SQLite to ProllyTree Import Tool
//!
//! This tool imports SQLite databases into ProllyTree for performance analysis.
//! It's designed to help identify performance bottlenecks in the tree implementation.
//!
//! Usage:
//!   cargo run --example sqlite_import --release -- <sqlite_file> [options]
//!
//! Examples:
//!   # In-memory with default node sizes
//!   cargo run --example sqlite_import --release -- mydata.db
//!
//!   # In-memory with large nodes
//!   cargo run --example sqlite_import --release -- mydata.db --node-size large
//!
//!   # With profiling info
//!   cargo run --example sqlite_import --release -- mydata.db --verbose

use prollytree::{
    config::TreeConfig,
    storage::InMemoryNodeStorage,
    tree::{ProllyTree, Tree},
};
use rusqlite::{Connection, Result as SqliteResult};
use std::collections::HashMap;
use std::env;
use std::time::Instant;

#[derive(Debug)]
struct TableInfo {
    columns: Vec<String>,
    primary_keys: Vec<String>,
}

fn get_all_tables(conn: &Connection) -> SqliteResult<Vec<String>> {
    let mut stmt = conn.prepare(
        "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'",
    )?;

    let tables = stmt
        .query_map([], |row| row.get(0))?
        .collect::<SqliteResult<Vec<String>>>()?;

    Ok(tables)
}

fn get_table_info(conn: &Connection, table_name: &str) -> SqliteResult<TableInfo> {
    let mut stmt = conn.prepare(&format!("PRAGMA table_info({})", table_name))?;

    let mut columns = Vec::new();
    let mut primary_keys = Vec::new();

    let rows = stmt.query_map([], |row| {
        Ok((
            row.get::<_, String>(1)?, // column name
            row.get::<_, i64>(5)?,    // pk flag
        ))
    })?;

    for row in rows {
        let (col_name, pk_flag) = row?;
        columns.push(col_name.clone());
        if pk_flag > 0 {
            primary_keys.push(col_name);
        }
    }

    // If no primary key, use rowid
    if primary_keys.is_empty() {
        primary_keys.push("rowid".to_string());
    }

    Ok(TableInfo {
        columns,
        primary_keys,
    })
}

fn build_primary_key(row_data: &HashMap<String, String>, pk_cols: &[String]) -> String {
    if pk_cols.len() == 1 {
        row_data.get(&pk_cols[0]).unwrap().clone()
    } else {
        pk_cols
            .iter()
            .map(|col| row_data.get(col).unwrap())
            .cloned()
            .collect::<Vec<_>>()
            .join(":")
    }
}

fn import_table(
    conn: &Connection,
    tree: &mut ProllyTree<32, InMemoryNodeStorage<32>>,
    table_name: &str,
    batch_size: usize,
    verbose: bool,
) -> SqliteResult<usize> {
    let info = get_table_info(conn, table_name)?;

    if verbose {
        println!("\nImporting table: {}", table_name);
        println!("  Columns: {}", info.columns.join(", "));
        println!("  Primary key(s): {}", info.primary_keys.join(", "));
    }

    // Build SELECT query
    let mut select_cols = info.columns.clone();
    if info.primary_keys.contains(&"rowid".to_string())
        && !info.columns.contains(&"rowid".to_string())
    {
        select_cols.insert(0, "rowid".to_string());
    }

    let query = format!("SELECT {} FROM {}", select_cols.join(", "), table_name);
    let mut stmt = conn.prepare(&query)?;

    let mut rows_iter = stmt.query([])?;
    let mut total_rows = 0;
    let mut batch_keys = Vec::new();
    let mut batch_values = Vec::new();

    let batch_start = Instant::now();

    while let Some(row) = rows_iter.next()? {
        // Build row data map
        let mut row_data = HashMap::new();
        for (i, col_name) in select_cols.iter().enumerate() {
            let value: String = row.get(i).unwrap_or_else(|_| "".to_string());
            row_data.insert(col_name.clone(), value);
        }

        // Build key: /<table>/<primary_key>
        let pk_value = build_primary_key(&row_data, &info.primary_keys);
        let key = format!("/{}/{}", table_name, pk_value);

        // Build value: JSON array of column values
        let value_list: Vec<String> = info
            .columns
            .iter()
            .map(|col| {
                row_data
                    .get(col)
                    .map(|v| format!("\"{}\"", v.replace('"', "\\\"")))
                    .unwrap_or_else(|| "null".to_string())
            })
            .collect();
        let value = format!("[{}]", value_list.join(","));

        batch_keys.push(key.into_bytes());
        batch_values.push(value.into_bytes());

        // Insert batch when it reaches batch_size
        if batch_keys.len() >= batch_size {
            tree.insert_batch(&batch_keys, &batch_values);
            total_rows += batch_keys.len();

            if verbose {
                let elapsed = batch_start.elapsed();
                let rate = total_rows as f64 / elapsed.as_secs_f64();
                let root_hash = tree.get_root_hash().unwrap_or_default();
                println!(
                    "  Inserted {} rows... ({:.1} rows/sec) Root: {:02x}{:02x}{:02x}{:02x}...",
                    total_rows,
                    rate,
                    root_hash.0[0],
                    root_hash.0[1],
                    root_hash.0[2],
                    root_hash.0[3]
                );
            }

            batch_keys.clear();
            batch_values.clear();
        }
    }

    // Insert remaining batch
    if !batch_keys.is_empty() {
        tree.insert_batch(&batch_keys, &batch_values);
        total_rows += batch_keys.len();
    }

    if verbose {
        let elapsed = batch_start.elapsed();
        let rate = total_rows as f64 / elapsed.as_secs_f64();
        println!(
            "  Inserted {} rows - DONE ({:.1} rows/sec)",
            total_rows, rate
        );
    }

    Ok(total_rows)
}

fn create_tree_config(node_size: &str) -> TreeConfig<32> {
    match node_size {
        "default" => TreeConfig::default(),
        "small" => TreeConfig {
            min_chunk_size: 100,
            max_chunk_size: 1_000_000,
            pattern: 0xFFFF, // ~1 in 65K split
            ..TreeConfig::default()
        },
        "medium" => TreeConfig {
            min_chunk_size: 500,
            max_chunk_size: 5_000_000,
            pattern: 0xFFFFF, // ~1 in 1M split
            ..TreeConfig::default()
        },
        "large" => TreeConfig {
            min_chunk_size: 1000,
            max_chunk_size: 10_000_000,
            pattern: 0xFFFFFF, // ~1 in 16M split
            ..TreeConfig::default()
        },
        "xlarge" => TreeConfig {
            min_chunk_size: 2000,
            max_chunk_size: 20_000_000,
            pattern: 0xFFFFFFF, // ~1 in 268M split
            ..TreeConfig::default()
        },
        _ => {
            eprintln!("Unknown node size: {}, using default", node_size);
            TreeConfig::default()
        }
    }
}

fn main() -> SqliteResult<()> {
    let args: Vec<String> = env::args().collect();

    if args.len() < 2 {
        eprintln!(
            "Usage: {} <sqlite_file> [--node-size SIZE] [--batch-size N] [--verbose]",
            args[0]
        );
        eprintln!("\nOptions:");
        eprintln!("  --node-size SIZE    Node size: default, small, medium, large, xlarge (default: default)");
        eprintln!("  --batch-size N      Batch size for inserts (default: 1000)");
        eprintln!("  --verbose           Print detailed progress");
        eprintln!("\nExamples:");
        eprintln!("  {} mydata.db", args[0]);
        eprintln!("  {} mydata.db --node-size large --verbose", args[0]);
        std::process::exit(1);
    }

    let sqlite_file = &args[1];
    let mut node_size = "default";
    let mut batch_size = 1000;
    let mut verbose = false;

    // Parse arguments
    let mut i = 2;
    while i < args.len() {
        match args[i].as_str() {
            "--node-size" => {
                if i + 1 < args.len() {
                    node_size = &args[i + 1];
                    i += 2;
                } else {
                    eprintln!("Error: --node-size requires a value");
                    std::process::exit(1);
                }
            }
            "--batch-size" => {
                if i + 1 < args.len() {
                    batch_size = args[i + 1].parse().unwrap_or(1000);
                    i += 2;
                } else {
                    eprintln!("Error: --batch-size requires a value");
                    std::process::exit(1);
                }
            }
            "--verbose" | "-v" => {
                verbose = true;
                i += 1;
            }
            _ => {
                eprintln!("Unknown argument: {}", args[i]);
                std::process::exit(1);
            }
        }
    }

    println!("Opening SQLite database: {}", sqlite_file);
    let conn = Connection::open(sqlite_file)?;

    println!("Getting table list...");
    let tables = get_all_tables(&conn)?;

    if tables.is_empty() {
        println!("No tables found in database");
        return Ok(());
    }

    println!("Tables to import: {}", tables.join(", "));

    println!("\nInitializing ProllyTree");
    println!("  Storage: in-memory");
    println!("  Node size: {}", node_size);
    println!("  Batch size: {}", batch_size);

    let config = create_tree_config(node_size);
    let storage = InMemoryNodeStorage::<32>::new();
    let mut tree = ProllyTree::new(storage, config);

    let total_start = Instant::now();
    let mut total_rows = 0;

    for table in &tables {
        match import_table(&conn, &mut tree, table, batch_size, verbose) {
            Ok(rows) => {
                total_rows += rows;
            }
            Err(e) => {
                eprintln!("Error importing table {}: {}", table, e);
            }
        }
    }

    let total_elapsed = total_start.elapsed();
    let overall_rate = total_rows as f64 / total_elapsed.as_secs_f64();

    println!("\n{}", "=".repeat(60));
    println!("Import complete!");
    println!("  Total rows imported: {}", total_rows);
    println!("  Total time: {:.2}s", total_elapsed.as_secs_f64());
    println!("  Overall rate: {:.1} rows/sec", overall_rate);
    println!("  Tree size: {}", tree.size());
    if let Some(root_hash) = tree.get_root_hash() {
        println!("  Root hash: {:x}", root_hash);
    }
    println!("{}", "=".repeat(60));

    Ok(())
}
